File: MonglePetTests/PetSpeechRuntimeTests.swift
import XCTest
@testable import MonglePet

@MainActor
final class PetSpeechRuntimeTests: XCTestCase {
    func testPeriodicSpeechUsesOneShotTimerAndReschedules() throws {
        let periodicScheduler = ManualPetSpeechScheduler()
        let dismissalScheduler = ManualPetSpeechScheduler()
        var presentations: [PetSpeechPresentation?] = []
        let phrase = PetSpeechPhrase(
            id: UUID(
                uuidString: "10000000-0000-0000-0000-000000000001"
            )!,
            text: "잠깐 쉬어 갈까요?",
            displayDurationMilliseconds: 4_000,
            trigger: .periodic
        )
        let theme = PetSpeechBubbleTheme(
            colorStyle: .midnight,
            fontSize: 16,
            showsTail: true,
            tailAlignment: .leading
        )
        let placement = PetSpeechBubblePlacementSettings(
            preferredPosition: .below,
            horizontalOffset: 32,
            gap: 14
        )
        let runtime = PetSpeechRuntime(
            scheduler: periodicScheduler,
            dismissalScheduler: dismissalScheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            presentations.append($0)
        }

        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                periodicIntervalMilliseconds: 15_000,
                phrases: [phrase],
                theme: theme,
                placement: placement
            )
        )
        XCTAssertNil(periodicScheduler.scheduledDelay)

        runtime.setAwake(true)
        XCTAssertEqual(periodicScheduler.scheduledDelay, .seconds(15))

        periodicScheduler.fire()

        XCTAssertEqual(
            presentations.compactMap { $0 }.last,
            PetSpeechPresentation(
                phraseID: phrase.id,
                text: phrase.text,
                displayDurationMilliseconds: 4_000,
                theme: theme,
                placement: placement
            )
        )
        XCTAssertNil(periodicScheduler.scheduledDelay)
        XCTAssertEqual(dismissalScheduler.scheduledDelay, .seconds(4))

        dismissalScheduler.fire()

        XCTAssertNil(presentations.last!)
        XCTAssertEqual(periodicScheduler.scheduledDelay, .seconds(15))
    }

    func testSequenceSpeechAppearsOnlyWhenSequenceChanges() {
        let scheduler = ManualPetSpeechScheduler()
        var receivedTexts: [String] = []
        let runtime = PetSpeechRuntime(
            scheduler: scheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            if let text = $0?.text {
                receivedTexts.append(text)
            }
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                phrases: [
                    PetSpeechPhrase(
                        text: "집중 시작!",
                        trigger: .sequence("focus")
                    ),
                    PetSpeechPhrase(
                        text: "쉬는 시간이에요.",
                        trigger: .sequence("rest")
                    )
                ]
            )
        )
        runtime.setAwake(true)

        runtime.behaviorSequenceDidChange("focus")
        runtime.behaviorSequenceDidChange("focus")
        runtime.behaviorSequenceDidChange("rest")

        XCTAssertEqual(receivedTexts, ["집중 시작!", "쉬는 시간이에요."])
        XCTAssertNil(scheduler.scheduledDelay)
    }

    func testPeriodicSpeechAvoidsImmediatelyRepeatingPhraseWhenPossible() {
        let scheduler = ManualPetSpeechScheduler()
        let first = PetSpeechPhrase(
            text: "첫 번째",
            trigger: .periodic,
            displayMode: .untilNextPhrase
        )
        let second = PetSpeechPhrase(
            text: "두 번째",
            trigger: .periodic,
            displayMode: .untilNextPhrase
        )
        var excludedIDs: [UUID?] = []
        var receivedTexts: [String] = []
        let runtime = PetSpeechRuntime(
            scheduler: scheduler,
            phrasePicker: { phrases, excludedID in
                excludedIDs.append(excludedID)
                return phrases.first { $0.id != excludedID }
            }
        ) {
            if let text = $0?.text {
                receivedTexts.append(text)
            }
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                periodicIntervalMilliseconds: 10_000,
                phrases: [first, second]
            )
        )
        runtime.setAwake(true)

        scheduler.fire()
        scheduler.fire()

        XCTAssertEqual(receivedTexts, ["첫 번째", "두 번째"])
        XCTAssertEqual(excludedIDs.count, 2)
        XCTAssertNil(excludedIDs[0])
        XCTAssertEqual(excludedIDs[1], first.id)
    }

    func testSleepAndSystemSuspensionCancelAndHideSpeech() {
        let scheduler = ManualPetSpeechScheduler()
        let dismissalScheduler = ManualPetSpeechScheduler()
        var presentations: [PetSpeechPresentation?] = []
        let runtime = PetSpeechRuntime(
            scheduler: scheduler,
            dismissalScheduler: dismissalScheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            presentations.append($0)
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                periodicIntervalMilliseconds: 10_000,
                phrases: [
                    PetSpeechPhrase(text: "안녕하세요!", trigger: .periodic)
                ]
            )
        )
        runtime.setAwake(true)
        XCTAssertEqual(scheduler.scheduledDelay, .seconds(10))
        scheduler.fire()
        XCTAssertNotNil(presentations.last!)
        XCTAssertEqual(dismissalScheduler.scheduledDelay, .seconds(3))

        runtime.setSystemSuspended(true)
        XCTAssertNil(scheduler.scheduledDelay)
        XCTAssertNil(dismissalScheduler.scheduledDelay)
        XCTAssertNil(presentations.last!)

        runtime.setSystemSuspended(false)
        XCTAssertEqual(scheduler.scheduledDelay, .seconds(10))

        runtime.setAwake(false)
        XCTAssertNil(scheduler.scheduledDelay)
        XCTAssertNil(dismissalScheduler.scheduledDelay)
    }

    func testDisabledSpeechDoesNotScheduleOrPresent() {
        let scheduler = ManualPetSpeechScheduler()
        var presentations: [PetSpeechPresentation?] = []
        let runtime = PetSpeechRuntime(
            scheduler: scheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            presentations.append($0)
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: false,
                phrases: [
                    PetSpeechPhrase(
                        text: "보이면 안 돼요.",
                        trigger: .sequence("focus")
                    )
                ]
            )
        )
        runtime.setAwake(true)
        runtime.behaviorSequenceDidChange("focus")

        XCTAssertNil(scheduler.scheduledDelay)
        XCTAssertTrue(presentations.compactMap { $0 }.isEmpty)
    }

    func testPetChangeDismissesSpeechEvenWhenNextSettingsAreEqual() {
        let periodicScheduler = ManualPetSpeechScheduler()
        let dismissalScheduler = ManualPetSpeechScheduler()
        var presentations: [PetSpeechPresentation?] = []
        let settings = PetSpeechSettings(
            isEnabled: true,
            periodicIsEnabled: true,
            periodicIntervalMilliseconds: 5_000,
            phrases: [
                PetSpeechPhrase(
                    text: "첫 번째 펫의 대사",
                    trigger: .periodic,
                    displayMode: .untilNextPhrase
                )
            ]
        )
        let runtime = PetSpeechRuntime(
            scheduler: periodicScheduler,
            dismissalScheduler: dismissalScheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            presentations.append($0)
        }
        runtime.update(settings: settings)
        runtime.setAwake(true)
        periodicScheduler.fire()
        XCTAssertNotNil(presentations.last!)

        runtime.prepareForPetChange()

        XCTAssertNil(presentations.last!)
        XCTAssertNil(periodicScheduler.scheduledDelay)
        XCTAssertNil(dismissalScheduler.scheduledDelay)

        runtime.update(settings: settings)
        runtime.setAwake(true)

        XCTAssertEqual(periodicScheduler.scheduledDelay, .seconds(5))
    }

    func testBehaviorSpeechImmediatelyReplacesPeriodicAndHasPriority() {
        let periodicScheduler = ManualPetSpeechScheduler()
        let dismissalScheduler = ManualPetSpeechScheduler()
        var receivedTexts: [String?] = []
        let runtime = PetSpeechRuntime(
            scheduler: periodicScheduler,
            dismissalScheduler: dismissalScheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            receivedTexts.append($0?.text)
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                periodicIsEnabled: true,
                periodicIntervalMilliseconds: 10_000,
                phrases: [
                    PetSpeechPhrase(
                        text: "주기 대사",
                        trigger: .periodic,
                        displayMode: .untilNextPhrase
                    ),
                    PetSpeechPhrase(
                        text: "집중 시작",
                        trigger: .sequence("focus"),
                        displayMode: .untilNextPhrase
                    )
                ]
            )
        )
        runtime.setAwake(true)
        periodicScheduler.fire()
        XCTAssertEqual(receivedTexts.last!, "주기 대사")

        runtime.behaviorSequenceDidChange("focus")

        XCTAssertEqual(receivedTexts.last!, "집중 시작")
        XCTAssertNil(periodicScheduler.scheduledDelay)
        periodicScheduler.fire()
        XCTAssertEqual(receivedTexts.last!, "집중 시작")
    }

    func testBehaviorChangeWithoutPhraseUsesDismissPolicy() {
        let periodicScheduler = ManualPetSpeechScheduler()
        var receivedTexts: [String?] = []
        let runtime = PetSpeechRuntime(
            scheduler: periodicScheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            receivedTexts.append($0?.text)
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                periodicIsEnabled: true,
                periodicIntervalMilliseconds: 8_000,
                behaviorChangePolicy: .dismiss,
                phrases: [
                    PetSpeechPhrase(
                        text: "집중 시작",
                        trigger: .sequence("focus"),
                        displayMode: .untilNextPhrase
                    ),
                    PetSpeechPhrase(
                        text: "주기 대사",
                        trigger: .periodic,
                        displayMode: .untilNextPhrase
                    )
                ]
            )
        )
        runtime.setAwake(true)
        runtime.behaviorSequenceDidChange("focus")

        runtime.behaviorSequenceDidChange("rest")

        XCTAssertNil(receivedTexts.last!)
        XCTAssertEqual(periodicScheduler.scheduledDelay, .seconds(8))
    }

    func testKeepPolicyRetainsBehaviorUntilPeriodicReplacement() {
        let periodicScheduler = ManualPetSpeechScheduler()
        var receivedTexts: [String?] = []
        let runtime = PetSpeechRuntime(
            scheduler: periodicScheduler,
            phrasePicker: { phrases, _ in phrases.first }
        ) {
            receivedTexts.append($0?.text)
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                periodicIsEnabled: true,
                periodicIntervalMilliseconds: 6_000,
                behaviorChangePolicy: .keep,
                phrases: [
                    PetSpeechPhrase(
                        text: "집중 시작",
                        trigger: .sequence("focus"),
                        displayMode: .untilNextPhrase
                    ),
                    PetSpeechPhrase(
                        text: "다음 이야기",
                        trigger: .periodic,
                        displayMode: .untilNextPhrase
                    )
                ]
            )
        )
        runtime.setAwake(true)
        runtime.behaviorSequenceDidChange("focus")

        runtime.behaviorSequenceDidChange("rest")

        XCTAssertEqual(receivedTexts.last!, "집중 시작")
        XCTAssertEqual(periodicScheduler.scheduledDelay, .seconds(6))

        periodicScheduler.fire()

        XCTAssertEqual(receivedTexts.last!, "다음 이야기")
    }

    func testSequentialPeriodicOrderWrapsToFirstPhrase() {
        let scheduler = ManualPetSpeechScheduler()
        var receivedTexts: [String] = []
        let phrases = ["하나", "둘", "셋"].map {
            PetSpeechPhrase(
                text: $0,
                trigger: .periodic,
                displayMode: .untilNextPhrase
            )
        }
        let runtime = PetSpeechRuntime(
            scheduler: scheduler,
            phrasePicker: { _, _ in
                XCTFail("순차 재생에서는 무작위 선택기를 사용하지 않아야 합니다.")
                return nil
            }
        ) {
            if let text = $0?.text {
                receivedTexts.append(text)
            }
        }
        runtime.update(
            settings: PetSpeechSettings(
                isEnabled: true,
                periodicIsEnabled: true,
                periodicIntervalMilliseconds: 5_000,
                periodicOrder: .sequential,
                phrases: phrases
            )
        )
        runtime.setAwake(true)

        scheduler.fire()
        scheduler.fire()
        scheduler.fire()
        scheduler.fire()

        XCTAssertEqual(receivedTexts, ["하나", "둘", "셋", "하나"])
    }
}

@MainActor
private final class ManualPetSpeechScheduler: PetSpeechScheduling {
    private(set) var scheduledDelay: Duration?
    private var action: (() -> Void)?

    func schedule(after delay: Duration, action: @escaping () -> Void) {
        scheduledDelay = delay
        self.action = action
    }

    func cancel() {
        scheduledDelay = nil
        action = nil
    }

    func fire() {
        let pendingAction = action
        scheduledDelay = nil
        action = nil
        pendingAction?()
    }
}

File: apps/windows/tests/MonglePet.Settings.Tests/PetSpeechRuntimeTests.cs
using MonglePet.Settings;

namespace MonglePet.Settings.Tests;

public sealed class PetSpeechRuntimeTests
{
    [Fact]
    public void PeriodicTimedPhraseWaitsThenDismissesAndReschedules()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase phrase = Phrase("안녕하세요", new PetSpeechTrigger.Periodic());
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);

        runtime.Update(Settings([phrase], periodic: true));
        runtime.SetAwake(true);

        Assert.Equal(TimeSpan.FromSeconds(60), periodic.Delay);
        periodic.Fire();
        Assert.Equal(phrase.Id, Assert.Single(presentations)!.PhraseId);
        Assert.Equal(TimeSpan.FromSeconds(3), dismissal.Delay);

        dismissal.Fire();
        Assert.Null(presentations[^1]);
        Assert.Equal(TimeSpan.FromSeconds(60), periodic.Delay);
    }

    [Fact]
    public void BehaviorPhrasePreemptsPeriodicAndSameSequenceDoesNotRepeat()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase periodicPhrase = Phrase("주기", new PetSpeechTrigger.Periodic());
        PetSpeechPhrase behaviorPhrase = Phrase("집중", new PetSpeechTrigger.Sequence("focus"));
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);
        runtime.Update(Settings([periodicPhrase, behaviorPhrase], periodic: true));
        runtime.SetAwake(true);

        periodic.Fire();
        runtime.BehaviorSequenceDidChange("focus");
        runtime.BehaviorSequenceDidChange("focus");

        Assert.Equal([periodicPhrase.Id, behaviorPhrase.Id], presentations
            .Where(value => value is not null)
            .Select(value => value!.PhraseId));
        Assert.False(periodic.IsScheduled);
    }

    [Fact]
    public void DismissPolicyClearsPhraseWhenBehaviorHasNoSpeech()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase phrase = Phrase(
            "집중",
            new PetSpeechTrigger.Sequence("focus"),
            PetSpeechDisplayMode.UntilNextPhrase);
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);
        runtime.Update(Settings([phrase]));
        runtime.SetAwake(true);
        runtime.BehaviorSequenceDidChange("focus");

        runtime.BehaviorSequenceDidChange("rest");

        Assert.Null(presentations[^1]);
    }

    [Fact]
    public void KeepPolicyRetainsPhraseAndSchedulesPeriodicReplacement()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase behavior = Phrase(
            "집중",
            new PetSpeechTrigger.Sequence("focus"),
            PetSpeechDisplayMode.UntilNextPhrase);
        PetSpeechPhrase periodicPhrase = Phrase(
            "쉬어요",
            new PetSpeechTrigger.Periodic(),
            PetSpeechDisplayMode.UntilNextPhrase);
        PetSpeechSettings settings = Settings([behavior, periodicPhrase], periodic: true) with
        {
            BehaviorChangePolicy = PetSpeechBehaviorChangePolicy.Keep,
        };
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);
        runtime.Update(settings);
        runtime.SetAwake(true);
        runtime.BehaviorSequenceDidChange("focus");

        runtime.BehaviorSequenceDidChange("rest");

        Assert.Equal(behavior.Id, presentations[^1]!.PhraseId);
        Assert.True(periodic.IsScheduled);
    }

    [Fact]
    public void SequentialPeriodicOrderWrapsAround()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase first = Phrase(
            "하나",
            new PetSpeechTrigger.Periodic(),
            PetSpeechDisplayMode.UntilNextPhrase);
        PetSpeechPhrase second = Phrase(
            "둘",
            new PetSpeechTrigger.Periodic(),
            PetSpeechDisplayMode.UntilNextPhrase);
        PetSpeechSettings settings = Settings([first, second], periodic: true) with
        {
            PeriodicOrder = PetSpeechPeriodicOrder.Sequential,
        };
        using var runtime = Runtime(periodic, dismissal, presentations);
        runtime.Update(settings);
        runtime.SetAwake(true);

        periodic.Fire();
        periodic.Fire();
        periodic.Fire();

        Assert.Equal([first.Id, second.Id, first.Id], presentations.Select(value => value!.PhraseId));
    }

    [Fact]
    public void BehaviorPhraseUntilNextSchedulesPeriodicReplacement()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase behavior = Phrase(
            "집중 중이에요",
            new PetSpeechTrigger.Sequence("focus"),
            PetSpeechDisplayMode.UntilNextPhrase);
        PetSpeechPhrase periodicPhrase = Phrase(
            "잠깐 쉬어도 좋아요",
            new PetSpeechTrigger.Periodic(),
            PetSpeechDisplayMode.UntilNextPhrase);
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);
        runtime.Update(Settings([behavior, periodicPhrase], periodic: true));
        runtime.SetAwake(true);

        runtime.BehaviorSequenceDidChange("focus");

        Assert.True(periodic.IsScheduled);
        Assert.Equal(TimeSpan.FromSeconds(60), periodic.Delay);
        periodic.Fire();
        Assert.Equal(periodicPhrase.Id, presentations[^1]!.PhraseId);
        Assert.True(periodic.IsScheduled);
    }

    [Fact]
    public void UpdatingPeriodicIntervalCancelsOldReservationAndUsesNewInterval()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase phrase = Phrase("안녕하세요", new PetSpeechTrigger.Periodic());
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);
        runtime.Update(Settings([phrase], periodic: true));
        runtime.SetAwake(true);

        runtime.Update(Settings([phrase], periodic: true) with
        {
            PeriodicIntervalMilliseconds = 5_000,
        });

        Assert.True(periodic.IsScheduled);
        Assert.Equal(TimeSpan.FromSeconds(5), periodic.Delay);
    }

    [Fact]
    public void SleepAndSystemSuspensionCancelAndHideSpeech()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase phrase = Phrase("안녕하세요", new PetSpeechTrigger.Periodic());
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);
        runtime.Update(Settings([phrase], periodic: true));
        runtime.SetAwake(true);
        periodic.Fire();

        runtime.SetSystemSuspended(true);

        Assert.Null(presentations[^1]);
        Assert.False(periodic.IsScheduled);
        Assert.False(dismissal.IsScheduled);

        runtime.SetSystemSuspended(false);
        Assert.True(periodic.IsScheduled);
        runtime.SetAwake(false);
        Assert.False(periodic.IsScheduled);
    }

    [Fact]
    public void PetChangeDismissesEvenWhenNextSettingsWouldMatch()
    {
        var periodic = new ManualScheduler();
        var dismissal = new ManualScheduler();
        var presentations = new List<PetSpeechPresentation?>();
        PetSpeechPhrase phrase = Phrase("안녕하세요", new PetSpeechTrigger.Periodic());
        using var runtime = Runtime(periodic, dismissal, presentations, (phrases, _) => phrases[0]);
        runtime.Update(Settings([phrase], periodic: true));
        runtime.SetAwake(true);
        periodic.Fire();

        runtime.PrepareForPetChange();

        Assert.Null(presentations[^1]);
        Assert.False(periodic.IsScheduled);
        Assert.False(dismissal.IsScheduled);
    }

    [Fact]
    public void PlacementUsesAboveThenFallsBelowAtTopEdge()
    {
        var visible = new PetSpeechBubbleRect(-1920, 0, 1920, 1080);
        var size = new PetSpeechBubbleSize(240, 80);
        var aboveParent = new PetSpeechBubbleRect(-1200, 400, 192, 192);
        var topParent = new PetSpeechBubbleRect(-1200, 10, 192, 192);

        PetSpeechBubblePlacementResult above = PetSpeechBubblePlacement.Calculate(
            aboveParent,
            size,
            visible,
            PetSpeechBubblePlacementSettings.Default);
        PetSpeechBubblePlacementResult below = PetSpeechBubblePlacement.Calculate(
            topParent,
            size,
            visible,
            PetSpeechBubblePlacementSettings.Default);

        Assert.Equal(PetSpeechBubbleTailEdge.Bottom, above.TailEdge);
        Assert.Equal(312, above.Origin.Y);
        Assert.Equal(PetSpeechBubbleTailEdge.Top, below.TailEdge);
        Assert.Equal(210, below.Origin.Y);
    }

    [Fact]
    public void AutomaticPlacementKeepsLockedSideForCurrentPresentation()
    {
        var visible = new PetSpeechBubbleRect(0, 0, 1000, 800);
        var size = new PetSpeechBubbleSize(240, 80);
        PetSpeechBubblePlacementResult initial = PetSpeechBubblePlacement.Calculate(
            new PetSpeechBubbleRect(400, 10, 100, 100),
            size,
            visible,
            PetSpeechBubblePlacementSettings.Default);
        PetSpeechBubblePlacementResult moved = PetSpeechBubblePlacement.Calculate(
            new PetSpeechBubbleRect(400, 400, 100, 100),
            size,
            visible,
            PetSpeechBubblePlacementSettings.Default,
            initial.TailEdge);

        Assert.Equal(PetSpeechBubbleTailEdge.Top, initial.TailEdge);
        Assert.Equal(PetSpeechBubbleTailEdge.Top, moved.TailEdge);
        Assert.Equal(508, moved.Origin.Y);
    }

    [Fact]
    public void PlacementClampsHorizontalOffsetAndPointsTailAtPet()
    {
        var parent = new PetSpeechBubbleRect(900, 500, 100, 100);
        var visible = new PetSpeechBubbleRect(0, 0, 1000, 800);
        var settings = PetSpeechBubblePlacementSettings.Default with
        {
            HorizontalOffset = 160,
        };

        PetSpeechBubblePlacementResult result = PetSpeechBubblePlacement.Calculate(
            parent,
            new PetSpeechBubbleSize(240, 80),
            visible,
            settings);

        Assert.Equal(760, result.Origin.X);
        Assert.Equal(190, result.TailAnchorX);
    }

    [Theory]
    [InlineData(0)]
    [InlineData(8)]
    [InlineData(64)]
    public void VisibleTailKeepsItsHeightAndMovesWithBubble(double gap)
    {
        PetSpeechBubbleTailMetrics metrics = PetSpeechBubbleTailLayout.Calculate(
            showsTail: true,
            baseTailHeight: 13,
            gap);

        Assert.Equal(13, metrics.TailHeight);
        Assert.Equal(gap, metrics.PlacementGap);
    }

    [Fact]
    public void HiddenTailKeepsGapAsEmptyPlacementSpace()
    {
        PetSpeechBubbleTailMetrics metrics = PetSpeechBubbleTailLayout.Calculate(
            showsTail: false,
            baseTailHeight: 13,
            gap: 24);

        Assert.Equal(0, metrics.TailHeight);
        Assert.Equal(24, metrics.PlacementGap);
    }

    private static PetSpeechRuntime Runtime(
        ManualScheduler periodic,
        ManualScheduler dismissal,
        List<PetSpeechPresentation?> presentations,
        PetSpeechRuntime.PhrasePicker? picker = null) =>
        new(periodic, dismissal, presentations.Add, picker);

    private static PetSpeechSettings Settings(
        IReadOnlyList<PetSpeechPhrase> phrases,
        bool periodic = false) =>
        new(
            true,
            periodic,
            60_000,
            PetSpeechPeriodicOrder.Random,
            PetSpeechBehaviorChangePolicy.Dismiss,
            phrases,
            PetSpeechBubbleTheme.Default,
            PetSpeechBubblePlacementSettings.Default);

    private static PetSpeechPhrase Phrase(
        string text,
        PetSpeechTrigger trigger,
        PetSpeechDisplayMode displayMode = PetSpeechDisplayMode.Timed) =>
        new(Guid.NewGuid(), text, 3_000, trigger, displayMode);

    private sealed class ManualScheduler : IPetSpeechScheduler
    {
        private Action? _action;

        public TimeSpan? Delay { get; private set; }

        public bool IsScheduled => _action is not null;

        public void Schedule(TimeSpan delay, Action action)
        {
            Delay = delay;
            _action = action;
        }

        public void Cancel()
        {
            Delay = null;
            _action = null;
        }

        public void Fire()
        {
            Action action = _action ?? throw new InvalidOperationException("예약된 작업이 없습니다.");
            _action = null;
            Delay = null;
            action();
        }
    }
}
